### DISCUSSION TITLE
absent_over_time should emit all the labels of that metric

### DISCUSSION THREAD

**Discussion by ankurnigam** [prometheus/prometheus#9374] [Ideas]
Hi Team, 

Absent and absent_over_time both are emitting only vector value but we expect absent_over_time should emit all the labels of the metric.

**Comment by alburthoffman** [comment]
what do u mean by emitting all labels?

**Comment by roidelapluie** [comment]
Hello, I understand your request, however, given the way Prometheus works, this is not possible. We only have access to instant data.

How would this even behave? would you query the entire history?

you can simulate what you want by using `unless` and `offset`:  `mymetric offset 1h unless mymetric` will, during one hour, showing disappearing labels.

**Reply by ankurnigam** [reply]
offset and unless can work for limited time , let's say if we put 1 hr as offset and alert remain there for two days , then resolved notification will get send after 1 hr.

**Reply by roidelapluie** [reply]
the alternative is to do `present_over_time(my_metric[7d])` unless my_metric. then you have a week but it will be very ressource consuming.

In general, it is recommended to avoid pushing to prometheus. if you push, you can use the pushgateway (if it is batches, see pushgateway's README). The pushgateway exposes the last timestamp he received the data as a metrics , so you could alert on that.

**Reply by ankurnigam** [reply]
correct. I was looking for something like absent by (instance) (my_metric). Also , I don't see present_over_time function , is this introduced recently?

**Reply by roidelapluie** [reply]
yes, in 2.29

**Reply by ankurnigam** [reply]
does it retain all labels of the metric?

**Reply by ankurnigam** [reply]
also do we have something else present as well in 2.29 ?

**Reply by roidelapluie** [reply]
yes it retains all the labels.

what do you mean with your second question?

**Comment by ankurnigam** [comment]
Actually our problem is that we have redesigned prometheus to use push mechanism instead of pull, so we are fetching metrics locally on the server and then we have another agent which is pushing metrics to prometheus . This is the standard cloud industry follows.

**Reply by RichiH** [reply]
Prometheus Remote Write is based on the assumption that the data goes through a Prometheus engine at some point between scraping and being pushed, it's not clearly defined otherwise.

I would suggest Cortex or Thanos as a long-term push target, instead.

**Reply by ankurnigam** [reply]
Even though we will have thanos but if function will not work then it would be a problem as absent over time doesn't give you labels , it just give you vector value so you cannot compare it dynamically.
